libbpf-tools: runqslower: add -c option to trace specified comm

chudihuang opened the pull request.

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -20,6 +20,8 @@ static volatile sig_atomic_t exiting = 0;
 struct env {
 	pid_t pid;
 	pid_t tid;
+	int comm_len;
```
How about using TASK_COMM_LEN instead of this variable?

chudihuang commented on libbpf-tools/runqslower.bpf.c:
```diff
@@ -26,8 +27,19 @@ struct {
+{
+        int i;
+
+        for (i = 0; i < TASK_COMM_LEN && targ_comm[i] != '\0'; i++) {
```
The comm_allowed function is designed to match processes whose command name prefix matches the string specified by the -c option. However, it appears that bpf_strncmp may not be well-suited for this particular prefix-matching scenario.

ekyooo replied:
Even in that case, I found that using bpf_strncmp works as shown below.
Sharing for reference.

```
+const volatile __u32 targ_comm_len = 0;
 const volatile __u64 min_us = 0;
 const volatile pid_t targ_pid = 0;
 const volatile pid_t targ_tgid = 0;
@@ -31,11 +32,17 @@ static __always_inline bool comm_allowed(const char *comm)
 {
        int i;

+       if (targ_comm_len)
+               return 0 == bpf_strncmp(comm, targ_comm_len, targ_comm);
+       else
+               return true;
```

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -33,18 +36,20 @@ const char *argp_program_bug_address =
 static const struct argp_option opts[] = {
 	{ "pid", 'p', "PID", 0, "Process PID to trace", 0 },
 	{ "tid", 't', "TID", 0, "Thread TID to trace", 0 },
+	{ "comm",  'c', "COMM",  0, "Trace this comm only", 0 },
```
If this patch is intended to "filter by process name prefix," it is not consistent with the option description.

ekyooo commented on libbpf-tools/runqslower.bpf.c:
```diff
@@ -26,8 +27,19 @@ struct {
+
 /* record enqueue timestamp */
-static int trace_enqueue(u32 tgid, u32 pid)
+static int trace_enqueue(u32 tgid, u32 pid,const char *comm)
```
Please always match the coding style of the base code.

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -20,6 +21,8 @@ static volatile sig_atomic_t exiting = 0;
 struct env {
 	pid_t pid;
 	pid_t tid;
+	int task_comm_len;
```
env can be seen as a kind of global variable, and I think it’s best to minimize its use unless absolutely necessary. In my opinion, task_comm_len doesn’t seem essential globally. 
If it’s still needed, at the very least, maintaining name consistency with comm seems necessary.

ekyooo reviewed (commented):
There were also comments regarding the title.

chudihuang commented:
@ekyooo
    Apologies for the oversight. I've thoroughly re-examined the changes and confirmed their correctness. Should any issues remain, please don't hesitate to point them out - I'll address them immediately.

ekyooo commented:
@chudihuang 
Two of my review comments did not receive your feedback. Could you provide feedback on them? They might affect the current code, so it would be good to proceed with the review after receiving your feedback.

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -169,6 +182,12 @@ int main(int argc, char **argv)
+	if (env.comm) {
+		size_t copy_len = strnlen(env.comm, TASK_COMM_LEN - 1);
+		memcpy((char*)obj->rodata->targ_comm, env.comm, copy_len);
+		obj->rodata->targ_comm[copy_len] = '\0';
```
How about changing the three lines to a single line like this?
```
snprintf(obj->rodata->targ_comm, TASK_COMM_LEN, "%s", env.comm);
```

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -9,6 +9,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
+#include <sys/param.h>
```
not needed.

chudihuang commented:
@ekyooo 
Thank you for the detailed code review. I have addressed the review comments and fixed the code. Please re-review:
1. Remove tailing whitespace. 
2. Consolidated three lines into a single statement as recommended.
3. Removed redundant #include <sys/param.h> declaration.

ekyooo commented:
@chudihuang 
There’s still some code that doesn’t follow the base coding style. How about finding and fixing it yourself to get the hang of it for future patches?

chudihuang commented:
@ekyooo 
Sorry for the formatting issues in the previous version. I've fixed two more style issues:

1. **Adjusted indentation** to use spaces consistently:  
   ```diff
   if (!comm_allowed(comm))
   -		return 0;
   +		return 0;  // Now using spaces for indentation

2. ​​Added missing spaces​​ after commas in function call:
- fprintf(stderr,"Warning: ...", env.comm,TASK_COMM_LEN-1, ...);
+ fprintf(stderr, "Warning: ...", env.comm, TASK_COMM_LEN - 1, ...);
Also fixed the line wrapping for better readability. Changes have been pushed. Please re-review!

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -169,6 +180,10 @@ int main(int argc, char **argv)
+	if (env.comm)
+		snprintf(obj->rodata->targ_comm, TASK_COMM_LEN, "%s", env.comm);
+
+
```
Remove unnecessary blank lines.

ekyooo commented on libbpf-tools/runqslower.c:
```diff
@@ -64,6 +68,13 @@ static error_t parse_arg(int key, char *arg, struct argp_state *state)
+	case 'c':
+		env.comm = strdup(arg);
+		len = strlen(arg);
+		if (len >= TASK_COMM_LEN)
```
If the len variable is only used once, is it really necessary to declare it? I'm curious about your thoughts.

ekyooo reviewed (commented):
Reviewers (not me) have to review a lot of patches. I think doing your best to reduce the reviewers' burden before requesting a review will help you get your review done.

chudihuang commented:
Sorry for the formatting issues. Perhaps we should consider implementing a tool like the Linux kernel's checkpatch.pl to automate code style validation?

chenhengqi commented on libbpf-tools/runqslower.bpf.c:
```diff
@@ -48,10 +62,13 @@ static int handle_switch(void *ctx, struct task_struct *prev, struct task_struct
 	struct event event = {};
 	u64 *tsp, delta_us;
 	u32 pid;
+	char comm[TASK_COMM_LEN] = {0};
```
We prefer `char comm[TASK_COMM_LEN] = {};`.

chenhengqi commented on libbpf-tools/runqslower.bpf.c:
```diff
@@ -99,13 +116,19 @@ int BPF_PROG(sched_switch, bool preempt, struct task_struct *prev, struct task_s
-	return trace_enqueue(BPF_CORE_READ(p, tgid), BPF_CORE_READ(p, pid));
+	char comm[TASK_COMM_LEN] = {0};
+
+	BPF_CORE_READ_STR_INTO(&comm, p, comm);
```
We have to do string copy even if `targ_comm` is not set.
Consider adding a feature flag, if not set, pass `NULL` instead.

chenhengqi reviewed (approved):
LGTM overall. But let's wait for input from @ekyooo .

ekyooo commented:
@chudihuang 
LGTM too, thanks for the effort.

> automate code style validation?

I think it's a good suggestion. If you're interested, it would be great if you could contribute.
Thank you.

@chenhengqi 
Thank you. :smile:
